### PR TITLE
Add entity to collection entity page

### PR DESCRIPTION
**Pull request by prabalsingh24**
### Problem
Add Entity To Collection From Entity Page

### Solution

1. Changed all entityPage to React Class
2. Added `Add To Collection` Modal
3. Created POST `collection/:collectionId/add` route

### CODE REVIEW

**Comment by prabalsingh24** [comment]
@MonkeyDo  Have a look at thiss.
Idk why the buttons are collapsing in the footer. They're still working though.

Screenshot_2020-06-26 Harry Potter (Work) – BookBrainz(1)

**Comment by prabalsingh24** [comment]
what do you think of `checkboxes` instead of numbers and `Add to selected collections` button next to cancel?

**Comment by prabalsingh24** [comment]
Screenshot_2020-06-27 I’m Just Here for More Food (Edition) – BookBrainz(1)

**Comment by prabalsingh24** [comment]
Instead of disabling the button, I did this.. We can do this for `merge` and `edit` too.. 

Screenshot_2020-06-27 Bitters (EditionGroup) – BookBrainz

**Comment by prabalsingh24** [comment]
Screenshot_2020-06-27 A A Long (Author) – BookBrainz(1)

Screenshot_2020-06-27 A A Long (Author) – BookBrainz(2)

**Comment by prabalsingh24** [comment]
Screenshot_2020-06-27 The New Moosewood Cookbook (Edition) – BookBrainz

**Comment by prabalsingh24** [comment]
what do you think of `large` size modal

File: src/client/components/pages/entities/footer.js [hunk 1]
```diff
@@ -18,72 +18,126 @@
...
+	onCloseModal() {
+		this.setState({showModal: false});
+	}
+
+	handleShowModal() {
+		if (this.props.user) {
+			this.setState({showModal: true});
+		}
+		else {
+			this.setState({error: 'You need to be logged in for this'});
```
Comment (MonkeyDo) [inline src/client/components/pages/entities/footer.js:53]:
```suggestion
			this.setState({error: 'You need to be logged in'});
```

File: src/client/components/pages/parts/add-to-collection-modal.js [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+	async componentDidUpdate(prevProps) {
+		const collections = await this.getCollections();
+		if (prevProps.show !== this.props.show) {
+			// eslint-disable-next-line react/no-did-update-set-state
+			this.setState({collectionsAvailable: collections, show: this.props.show});
+		}
+	}
+
+	toggleRow(collectionId) {
+		// eslint-disable-next-line react/no-access-state-in-setstate
```
Comment (MonkeyDo) [inline src/client/components/pages/parts/add-to-collection-modal.js:46]: That does not look good. Why is this needed?

File: src/client/components/pages/parts/add-to-collection-modal.js [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+		this.handleShowCollectionForm = this.handleShowCollectionForm.bind(this);
+		this.handleShowAllCollections = this.handleShowAllCollections.bind(this);
+		this.handleAddToNewCollection = this.handleAddToNewCollection.bind(this);
+		this.isValid = this.isValid.bind(this);
+	}
+
+	async componentDidUpdate(prevProps) {
+		const collections = await this.getCollections();
+		if (prevProps.show !== this.props.show) {
+			// eslint-disable-next-line react/no-did-update-set-state
```
Comment (MonkeyDo) [inline src/client/components/pages/parts/add-to-collection-modal.js:40]: Not 100% sure what's going on here.
Why do you need to store `show` on the state, instead of using `this.props.show` directly?

As for setting collectionsAvailable, I think it would be better to do it in `componentDidMount` instead. I don't really think we need to re-fetch if a user closes and reopens the modal.

File: src/client/components/pages/parts/add-to-collection-modal.js [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+			newSelected = oldSelected.push(collectionId) && oldSelected;
+		}
+		this.setState({
+			selectedCollections: newSelected
+		});
+	}
+
+	async getCollections() {
+		// Get all collections of the user (unlikely that a user will have more than 10000 collections
+		const req = await request.get(`/editor/${this.props.userId}/collections/collections?type=${this.props.entityType}&size=10000`);
```
Comment (MonkeyDo) [inline src/client/components/pages/parts/add-to-collection-modal.js:62]: Why the `size` here? Is it because we have a small default size?
If so, we might want to think about implementing `size=0` to fetch all the collections?

Your comment made me think about what would happen if there are a lot of collections.
Maybe at the end of the project we can think about using a virtualized scroll.

File: src/client/components/pages/parts/add-to-collection-modal.js [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+		const collections = req.body;
+		return collections;
+	}
+
+	handleClose() {
+		this.setState({show: false}, this.props.closeCallback);
+	}
```
Comment (MonkeyDo) [inline src/client/components/pages/parts/add-to-collection-modal.js:69]: To me it would make more sense to just call `onCloseCallback` and letting the parent component set this.props.show and using that directly.

File: src/client/components/pages/parts/add-to-collection-modal.js (hunk 3 above)
Comment (prabalsingh24) [reply src/client/components/pages/parts/add-to-collection-modal.js:40]: Yeah we can use `componentDidMount`. I was using `componentDidUpdate` because earlier I didn't have the feature of creating the collection in the modal. and had to open a new tab to create a new collection. In that case `componentDidUpdate` was required.
But now that we can create collection in the modal itself we don't need to fetch collections everytime. One time is enough

File: src/client/stylesheets/style.less [hunk 6]
```diff
@@ -478,3 +478,13 @@ hr.wide {
 .badge.new {
 	background-color: @green;
 }
+
+.label-checkbox {
+	padding: 0 .5em;
+	vertical-align: .25em;
```
[the remark this answers is not in the document]
Comment (prabalsingh24) [reply src/client/stylesheets/style.less:484]: This is how it is looking for `vertical-align: text-top;`

Screenshot_2020-06-30 W Pustyni i w Puszczy (Work) – BookBrainz

File: src/client/components/pages/parts/add-to-collection-modal.js [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+						this.state.showCollectionForm ?
+							<Button bsStyle="primary" onClick={this.handleAddToNewCollection}>
+								<FontAwesomeIcon icon="plus"/> Add to this new collection
+							</Button> :
+							<Button bsStyle="primary" onClick={this.handleAddToCollection}>
+								<FontAwesomeIcon icon="plus"/> Add to selected collections
+							</Button>
+					}
+					<Button bsStyle="danger" onClick={this.handleClose}>
+						<FontAwesomeIcon icon="times-circle"/> Cancel
```
[the remark this answers is not in the document]
Comment (prabalsingh24) [reply src/client/components/pages/parts/add-to-collection-modal.js:291]: left or right?

File: src/client/components/pages/parts/add-to-collection-modal.js [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+							)))
+						}
+					</div>
+				</div>
+			);
+		}
+		else {
+			existingCollections = (
+				<div>
+					Oops, Looks like you do not have any collection of type {this.props.entityType} . <br/>
```
[the remark this answers is not in the document]
Comment (prabalsingh24) [reply src/client/components/pages/parts/add-to-collection-modal.js:201]: Umm we don't really need to do this `entityType` is already `EditionGroup`

File: src/client/components/pages/parts/add-to-collection-modal.js [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+					message: {
+						text: 'Internal Error',
+						type: 'danger'
+					}
+				});
+			});
+	}
+
+	isValid() {
+		return this.name.getValue() && this.privacy.getValue();
```
[the remark this answers is not in the document]
Comment (prabalsingh24) [reply src/client/components/pages/parts/add-to-collection-modal.js:157]: Hmmm I'll make this change in `userCollectionForm` as well

File: src/server/routes/collection.js [hunk 10]
```diff
@@ -85,4 +84,28 @@ router.get('/:collectionId', (req, res) => {
...
+		const {UserCollectionItem} = req.app.locals.orm;
+		for (const bbid of bbids) {
+			try {
+				await new UserCollectionItem({
+					bbid,
+					collectionId: collection.id
+				}).save(null, {method: 'insert'});
+			}
+			catch (err) {
+				// entity already in the collection
```
[the remark this answers is not in the document]
Comment (prabalsingh24) [reply src/server/routes/collection.js:101]: Umm I don't really think we need to log this. Because user can accidentally add same entity multiple times and it's not really a bug. 
It would've been nice if bookshelf had `upsert` option.
The other thing we can do is first check if exists, if not then add it. But that would require 2 calls. And since this is gonna be used alot, I don't think we want this.

File: src/client/components/pages/parts/add-to-collection-modal.js [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+			entityType,
+			name,
+			privacy
+		};
+		const {bbids} = this.props;
+		request.post('/collection/create/handler')
+			.send(data)
+			.then((res) => {
+				request.post(`/collection/${res.body.id}/add`)
+					.send({bbids}).then(() => {
```
[the remark this answers is not in the document]
Comment (prabalsingh24) [reply src/client/components/pages/parts/add-to-collection-modal.js:138]: Yeah we can refactor it later..

File: src/client/components/pages/parts/add-to-collection-modal.js (hunk 4 above)
Comment (prabalsingh24) [reply src/client/components/pages/parts/add-to-collection-modal.js:62]: Yes it has default 20 size and also that endpoint uses `fetchPage` and we need to pass a size in it, otherwise it takes up default `10` size.

The modal is scrollable, so if there are alot of collectoins, the modal will become very long...

File: src/client/components/pages/parts/add-to-collection-modal.js (hunk 2 above)
Comment (prabalsingh24) [reply src/client/components/pages/parts/add-to-collection-modal.js:46]: Yeah this is not needed. I forgot to remove it

File: src/client/components/pages/parts/add-to-collection-modal.js (hunk 4 above)
Comment (MonkeyDo) [reply src/client/components/pages/parts/add-to-collection-modal.js:58]: OK, let's cross that bridge when we come to it, then.
There is a virtualized scrolling library already installed that will help us with that (we use it for the long list of languages to select from)

File: src/server/routes/collection.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (MonkeyDo) [reply src/server/routes/collection.js:101]: Yes, upsert would be useful here, definitely.
Let's not do the double call, as therse are no dire consequences if adding an entity twice.
The logging I meant more in case there is another type of error.

This can be achieved by checking the type of error, displaying an "entity already added" message if it's a [NoRowsUpdatedError](https://bookshelfjs.org/api.html#Model-static-NoRowsUpdatedError) and logging the error otherwise.

File: src/client/components/pages/parts/add-to-collection-modal.js [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+						this.setState({
+							message: {
+								text: `Successfully added to your new collection: ${name}`,
+								type: 'success'
+							}
+						});
```
Comment (MonkeyDo) [inline src/client/components/pages/parts/add-to-collection-modal.js:136]: This looks nice currently, but a user is allowed to continue clicking on "add to new collection" and it will create as many duplicate collections.
We can avoid that easily by either:
1. closing the modal (and maybe showing a success alert in the entity footer (but what about adding from the search page?) or 
2. disabling the button after successful creation. I think that is more elegant and self-contained.

File: src/client/components/pages/parts/add-to-collection-modal.js [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+					});
+			}, (error) => {
+				this.setState({
+					message: {
+						text: 'Internal Error: Try again later',
```
Comment (MonkeyDo) [inline src/client/components/pages/parts/add-to-collection-modal.js:141]:
```suggestion
						text: 'Something went wrong! Please try again later',
```

File: src/client/components/pages/parts/add-to-collection-modal.js [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+		this.setState({message: {}, showCollectionForm: false});
+	}
+
+	handleAddToNewCollection(evt) {
+		evt.preventDefault();
+
+		if (!this.isValid()) {
+			this.setState({
+				message: {
+					text: 'Internal Error: Try again later',
```
Comment (MonkeyDo) [inline src/client/components/pages/parts/add-to-collection-modal.js:107]: I liked the "Incomplete form"  message you had previously.
How about "The form is incomplete. Please fill in a name and privacy option before continuing."

File: src/client/components/pages/parts/add-to-collection-modal.js [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+						this.state.showCollectionForm ?
+							<Button bsStyle="success" onClick={this.handleAddToNewCollection}>
+								<FontAwesomeIcon icon="plus"/> Add to new collection
+							</Button> :
+							<Button bsStyle="success" onClick={this.handleAddToCollection}>
+								<FontAwesomeIcon icon="plus"/>Add to selected collection{this.state.selectedCollections.length > 1 ? 's' : null}
+							</Button>
+					}
+					<Button bsStyle="danger" onClick={this.props.onCloseCallback}>
+						<FontAwesomeIcon icon="times"/> Cancel
```
Comment (MonkeyDo) [inline src/client/components/pages/parts/add-to-collection-modal.js:278]: After some thought and using the modal, I feel like "Close" would be better than "Cancel" since if I added to a or created a collection, the button will not revert that action.
```suggestion
						<FontAwesomeIcon icon="times"/> Close
```


If I close the modal, the collection selection is not cleared (which is good IMO) but the new collection form is cleared. It's because when we toggle the collection rows it's stored in the state, while the name/description/privacy isn't.
I don't think that's an issue in this case, but something to keep in mind I guess.

File: src/client/components/pages/parts/add-to-collection-modal.js [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+			newSelected = [...oldSelected, collectionId];
+		}
+		this.setState({
+			selectedCollections: newSelected
+		});
+	}
+
+	async getCollections() {
+		// Get all collections of the user (unlikely that a user will have more than 10000 collections
+		const req = await request.get(`/editor/${this.props.userId}/collections/collections?type=${this.props.entityType}&size=10000`);
```
Comment (MonkeyDo) [inline src/client/components/pages/parts/add-to-collection-modal.js:58]: It might be a good idea to wrap this in a try/catch block for safety, and display a generic user-friendly error message if there is a failure, something like "Sorry, we could not fetch your collections ({error})"

File: src/client/components/pages/parts/add-to-collection-modal.js [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+		const collections = req.body;
+		return collections;
```
Comment (MonkeyDo) [inline src/client/components/pages/parts/add-to-collection-modal.js:60]:
```suggestion
		return req.body;
```

File: src/client/components/pages/parts/add-to-collection-modal.js [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+				message: {
+					text: `Successfully added to selected collection${selectedCollections.length > 1 ? 's' : ''}`,
+					type: 'success'
+				}
+			});
+		}
+		else {
+			this.setState({
+				message: {
+					text: 'No Collection Selected',
```
Comment (MonkeyDo) [inline src/client/components/pages/parts/add-to-collection-modal.js:86]:
```suggestion
					text: 'No collection selected',
```

File: src/client/components/pages/parts/add-to-collection-modal.js [hunk 19]
```diff
@@ -0,0 +1,296 @@
...
+						<FontAwesomeIcon icon="times"/> Cancel
+					</Button>
+				</Modal.Footer>
+			</Modal>
+		);
+	}
+}
+
+
+AddToCollectionModal.displayName = 'DeleteCollectionModal';
```
Comment (MonkeyDo) [inline src/client/components/pages/parts/add-to-collection-modal.js:287]:
```suggestion
AddToCollectionModal.displayName = 'AddToCollectionModal';
```

File: src/client/components/pages/parts/add-to-collection-modal.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (MonkeyDo) [reply src/client/components/pages/parts/add-to-collection-modal.js]: Right, it should be `Edition Groups` instead with a space

File: src/client/components/pages/parts/add-to-collection-modal.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (MonkeyDo) [reply src/client/components/pages/parts/add-to-collection-modal.js]: Hm, I think we may have inconsistent placement and color across the website. I was looking at the relationship editor where the close button is on the left.

**Comment by prabalsingh24** [comment]
> Maybe `Add to collection`?
> What do you think?

With `add to collection` the button is becoming bigger than others...

**Comment by prabalsingh24** [comment]
> This looks nice currently, but a user is allowed to continue clicking on "add to new collection" and it will create as many duplicate collections.
> We can avoid that easily by either:
> 
>     1. closing the modal (and maybe showing a success alert in the entity footer (but what about adding from the search page?) or
> 
>     2. disabling the button after successful creation. I think that is more elegant and self-contained.


<br><br>
I like the first option better. User might want to add to different collection too.

Screenshot_2020-06-30 Flatbreads Flavors (Edition) – BookBrainz(1)

<br>
As for search result page, I am planning to show error message like this, we can show `success` message too

Screenshot_2020-06-30 Search Results – BookBrainz

File: src/client/components/pages/parts/add-to-collection-modal.js [hunk 20]
```diff
@@ -0,0 +1,307 @@
...
+								<FontAwesomeIcon icon="plus"/>
+								&nbsp;New collection
+							</Button>
+					}
+					{
+						this.state.showCollectionForm ?
+							<Button bsStyle="success" onClick={this.handleAddToNewCollection}>
+								<FontAwesomeIcon icon="plus"/> Add to new collection
+							</Button> :
+							<Button bsStyle="success" onClick={this.handleAddToCollection}>
```
Comment (MonkeyDo) [inline src/client/components/pages/parts/add-to-collection-modal.js:283]: I'd suggest making this `disabled` if `!this.state.collectionsAvailable.length`

File: src/client/components/pages/parts/add-to-collection-modal.js [hunk 21]
```diff
@@ -0,0 +1,307 @@
...
+
+		return (
+			<Modal
+				scrollable
+				show={this.props.show}
+				onHide={this.props.handleCloseModal}
+			>
+				<Modal.Header closeButton>
+					<Modal.Title>
+						Select Collection
```
Comment (MonkeyDo) [inline src/client/components/pages/parts/add-to-collection-modal.js:252]:
```suggestion
						{this.state. showCollectionForm ? 'Create' : 'Select'} Collection
```

File: src/client/components/pages/parts/add-to-collection-modal.js [hunk 22]
```diff
@@ -0,0 +1,307 @@
...
+		if (selectedCollections.length) {
+			const promiseArray = [];
+			selectedCollections.forEach((collectionId) => {
+				const submissionURL = `/collection/${collectionId}/add`;
+				promiseArray.push(
+					request.post(submissionURL)
+						.send({bbids})
+				);
+			});
+			await Promise.all(promiseArray);
```
Comment (MonkeyDo) [inline src/client/components/pages/parts/add-to-collection-modal.js:85]: We probably want to wrap this in a try/catch block and display a generic error message if something goes south

File: src/server/routes/collection.js [hunk 23]
```diff
@@ -85,4 +85,31 @@ router.get('/:collectionId', (req, res) => {
...
+		const {UserCollectionItem} = req.app.locals.orm;
+		for (const bbid of bbids) {
+			try {
+				await new UserCollectionItem({
+					bbid,
+					collectionId: collection.id
+				}).save(null, {method: 'insert'});
+			}
+			catch (err) {
+				if (err.constraint !== 'user_collection_item_pkey') {
```
Comment (MonkeyDo) [inline src/server/routes/collection.js:102]: I wish there was a better way to check if it's an "entry already exists" error, but nothing I tried is better than this. So be it.
A comment would be useful above this line to explain why we're doing this, as it's not clear unless you're very familiar with the db errors (which I'll admit I'm not).

However in this block you would `throw err` instead of logging it so that it can be dealt with by the parent catch block. Otherwise you're in a situation where the addition failed but you're sending a success response.
